### PR TITLE
Implement P2438R2 `string::substr() &&`

### LINKED ISSUE #2928: P2438R2 `string::substr() &&`
[P2438R2](https://wg21.link/P2438R2) `string::substr() &&`

This paper intentionally doesn't provide a feature-test macro.

Please also ensure that the new `substr` rvalue overload and the pre-existing lvalue overload both correctly use a default-constructed allocator for the returned `string`. (See #3022)

### LINKED ISSUE #3022: `<string>`: Should `string::substr` forward the allocator to the newly created `string`?
The C++ Standard doesn't say that: https://eel.is/c++draft/string.substr

https://github.com/microsoft/STL/blob/5aae6780236208577b4e8a000cbe7aa3cbc24c41/stl/inc/xstring#L4673-L4676

@CaseyCarter agreed. He also assumed that we'll fix this when implementing WG21-P2438 but I still decided to create an issue to not forget about it.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
And make the old and new `substr` use default constructed allocators.

(And consistently refer to WG21-N4910 in `<xstring>`.)

Fixes #2928.
Fixes #3022.

As implementation details, the rvalue substring constructors reuse the source storage if and only if
- the provided allocator and the allocator of the source compare equal, and
- the constructed string doesn't fit into SSO buffer.

I have no idea how to portably test the probably intended behavior of newly added overloads...

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
It seems that EDG is buggy here:
- EDG's `__builtin_memcpy` seemly doesn't work well with constexpr-dynamically-allocated storage,
- EDG seemly mishandled overload resolution between `T::operator const U&` and `T::operator U&&`.

I'll skip relative cases for EDG.

File: tests/std/tests/GH_003022_substr_allocator/test.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        {STATICALLY_WIDEN(CharT, "Take a longer substring from a long string"), 20, 13},
+        {STATICALLY_WIDEN(CharT, "Take a quiet longer substring..."), 10, 20},
+        {STATICALLY_WIDEN(CharT, "Take the whole long string as a substring"), 0, 41},
+    };
+
+    using string_type = basic_string<CharT, char_traits<CharT>, payloaded_allocator<CharT>>;
+
+    auto myator = payloaded_allocator<CharT>{42};
+
+    for (const auto& test_case : substring_test_cases) {
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_003022_substr_allocator/test.cpp:138]: Suggestion: Extracting `offset` and `count` variables would improve the readability of the assertions, several of which are wrapped.

File: tests/std/tests/GH_003022_substr_allocator/test.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+        operator string_type&&() {
+            return move(str);
+        }
+    };
+
+#if _HAS_CXX23 && !defined(__EDG__) // TRANSITION, EDG doesn't perform overload resolution correctly
+    static_assert(!is_constructible_v<string_type, ambiguator&, size_t, size_t>, "Ambiguous");
+#else
+    static_assert(is_constructible_v<string_type, ambiguator&, size_t, size_t>, "Not ambiguous");
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_003022_substr_allocator/test.cpp:180]: Suggestion: These messages are slightly confusing as to whether they're describing what *should* happen, or what actually happened if they have failed.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks, this looks great! I'll validate and push changes for the minor issues and nitpicks I found.

File: tests/std/tests/GH_003022_substr_allocator/test.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        operator const string_type&() {
+            return str;
+        }
+
+        operator string_type&&() {
+            return move(str);
+        }
+    };
+
+#if _HAS_CXX23 && !defined(__EDG__) // TRANSITION, EDG doesn't perform overload resolution correctly
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply tests/std/tests/GH_003022_substr_allocator/test.cpp:177]: Reported as VSO-1601179 "EDG's overload resolution incorrectly considers some ambiguous conversions to be unambiguous" with a fully reduced test case:

<details>
<summary>Click to expand repro:</summary>

```
C:\Temp>"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvars64.bat"
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.4.0-pre.1.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Temp>type meow.cpp
```
```cpp
#include <type_traits>

struct String {
    String() = default;

    String(const String&, int, int) {}
    String(String&&, int, int) {}
};

struct Ambiguator {
    operator const String&();
    operator String&&();
};

static_assert(!std::is_constructible_v<String, Ambiguator&, int, int>, "This should be ambiguous");

int main() {}
```
```
C:\Temp>cl /EHsc /nologo /W4 /MT /std:c++latest /c meow.cpp
meow.cpp

C:\Temp>clang-cl /EHsc /nologo /W4 /MT /std:c++latest /c meow.cpp

C:\Temp>cl /EHsc /nologo /W4 /MT /std:c++latest /c /BE meow.cpp
meow.cpp
"meow.cpp", line 15: error: static assertion failed with "This should be
          ambiguous"
  static_assert(!std::is_constructible_v<String, Ambiguator&, int, int>, "This should be ambiguous");
  ^
```
</details>

**Review by CaseyCarter — APPROVED** [review]
Going to approve because I don't want to reset testing, but I'll address these comments in a followup.

File: stl/inc/xstring [hunk 4]
```diff
@@ -2629,6 +2646,19 @@ public:
             _Right._Mypair._Myval2._Myptr() + _Roff, _Right._Mypair._Myval2._Clamp_suffix_size(_Roff, _Count));
     }
 
+#if _HAS_CXX23
+    constexpr basic_string(basic_string&& _Right, const size_type _Roff, const _Alloc& _Al = _Alloc())
+        : _Mypair(_One_then_variadic_args_t{}, _Al) { // construct from _Right [_Roff, <end>), potentially move
+        _Move_construct_from_substr(_Right, _Roff, npos, _Al);
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/xstring:2652]: Good catch; I was under the impression that there were other calls, but there aren't.

**Comment by StephanTLavavej** [comment]
I've resolved a trivial adjacent-add conflict in `tests/std/test.lst`.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this feature in one of the STL's most-used data structures! :heart_eyes_cat: :white_check_mark: :yarn:

**Pull request merged**